src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1861,156 +1819,155 @@ final void registerWorker(WorkQueue w) {
+            if ((lockRunState() & STOP) != 0L)
+                replaceable = false;
+            else if ((qs = queues) != null && (n = qs.length) > 0 &&
+                     qs[i = phase & SMASK & (n - 1)] == w) {
                 qs[i] = null;
                 stealCount += ns;         // accumulate steals
             }
             unlockRunState();
+            if (replaceable)
+                signalWork();
         }
-        if ((runState & STOP) == 0 && replaceable)
-            signalWork(null, 0); // may replace unless trimmed or uninitialized
         if (ex != null)
             ForkJoinTask.rethrow(ex);
     }
 
     /**
-     * Releases an idle worker, or creates one if not enough exist,
-     * returning on contention if a signal task is already taken.
-     *
-     * @param a if nonnull, a task array holding task signalled
-     * @param k index of task in array
+     * Releases an idle worker, or creates one if not enough exist.
      */
-    final void signalWork(ForkJoinTask<?>[] a, int k) {
+    final void signalWork() {
         int pc = parallelism;
         for (long c = ctl;;) {
             WorkQueue[] qs = queues;
             long ac = (c + RC_UNIT) & RC_MASK, nc;
             int sp = (int)c, i = sp & SMASK;
-            if (qs == null || qs.length <= i)
+            if ((short)(c >>> RC_SHIFT) >= pc)
+                break;
+            if (qs == null)
+                break;
+            if (qs.length <= i)
                 break;
             WorkQueue w = qs[i], v = null;
             if (sp == 0) {
                 if ((short)(c >>> TC_SHIFT) >= pc)
                     break;
                 nc = ((c + TC_UNIT) & TC_MASK);
             }
-            else if ((short)(c >>> RC_SHIFT) >= pc || (v = w) == null)
+            else if ((v = w) == null)
                 break;
             else
                 nc = (v.stackPred & LMASK) | (c & TC_MASK);
-            if (c == (c = compareAndExchangeCtl(c, nc | ac))) {
+            if (c == (c = ctl) && c == (c = compareAndExchangeCtl(c, nc | ac))) {
                 if (v == null)
                     createWorker();
                 else {
                     v.phase = sp;
-                    if (v.parking != 0)
+                    if (v.source < 0)
                         U.unpark(v.owner);
                 }
                 break;
             }
-            if (a != null && k >= 0 && k < a.length && a[k] == null)
+            if ((runState & STOP) != 0L)
                 break;
         }
     }
 
     /**
-     * Reactivates the given worker, and possibly others if not top of
-     * ctl stack. Called only during shutdown to ensure release on
-     * termination.
+     * Releases all waiting workers. Called only during shutdown.
+     *
+     * @return current ctl
      */
-    private void reactivate(WorkQueue w) {
-        for (long c = ctl;;) {
+    private long releaseAll() {
+        long c = ctl;
```

Comment:
`ctl` is not re-read inside the endless loop, is there any risk for a stale `ctl` to make `releaseAll()` never terminate?

Reply:
It is implicitly read via the compareAndExchangeCtl